tests/utils/test_log_handlers.py

```diff
@@ -313,28 +313,36 @@ def test__read_for_k8s_executor(self, mock_k8s_get_task_log, create_task_instanc
         else:
             mock_k8s_get_task_log.assert_not_called()
 
-    def test__read_for_celery_executor_fallbacks_to_worker(self, create_task_instance):
+    @pytest.mark.parametrize(
+        "state", [TaskInstanceState.RUNNING, TaskInstanceState.DEFERRED, TaskInstanceState.UP_FOR_RETRY]
+    )
+    def test__read_for_celery_executor_fallbacks_to_worker(self, state, create_task_instance):
         """Test for executors which do not have `get_task_log` method, it fallbacks to reading
         log from worker if and only if remote logs aren't found"""
         executor_name = "CeleryExecutor"
-
+        # Reading logs from worker should occur when the task is either running, deferred, or up for retry.
         ti = create_task_instance(
-            dag_id="dag_for_testing_celery_executor_log_read",
+            dag_id=f"dag_for_testing_celery_executor_log_read_{state}",
             task_id="task_for_testing_celery_executor_log_read",
             run_type=DagRunType.SCHEDULED,
             execution_date=DEFAULT_DATE,
         )
-        ti.state = TaskInstanceState.RUNNING
         ti.try_number = 2
+        ti.state = state
         with conf_vars({("core", "executor"): executor_name}):
             reload(executor_loader)
             fth = FileTaskHandler("")
-
             fth._read_from_logs_server = mock.Mock()
             fth._read_from_logs_server.return_value = ["this message"], ["this\nlog\ncontent"]
             actual = fth._read(ti=ti, try_number=2)
             fth._read_from_logs_server.assert_called_once()
-            assert actual == ("*** this message\nthis\nlog\ncontent", {"end_of_log": False, "log_pos": 16})
+            # If we are in the up for retry or deferred state, log has ended and a new task try has started.
+            # When the state is running, the log has not ended.
+            expected_end_of_log = state in (TaskInstanceState.DEFERRED, TaskInstanceState.UP_FOR_RETRY)
```

Comment:
I would just inline this since you don't seem to use it anywhere else, but I won't block for that.

Reply:
Hmm, inlining it would make the comment explaining the logic be placed in a somewhat awkward position imo, or that's why I think I wrote it like this.

Reply:
Yeah, that's fair.  Feel free to just resolve this convo.